### DISCUSSION TITLE
Start trajectory synchronously for several crazyfile

### DISCUSSION THREAD

**Discussion by CYDXYYJ** [USC-ACTLab/crazyswarm#438] [Q&A]
I try to start trajectory synchronously for several crazyfile by this code:
```
    swarm = Crazyswarm()
    allcfs = swarm.allcfs

    traj = uav_trajectory.Trajectory()
    
    for i in range(Num):
        traj.input_trajectory(agent[i].traj)
        allcfs.crazyflies[i].uploadTrajectory(0, 0, traj)

    allcfs.startTrajectory(0,relative=False)
```
But I find  they ran a false trajectory even if they did not run the same trajectory.

However, I used following code and they ran the correct trajectory even if it is not synchronous.
```
    for i in range(Num):
        traj.input_trajectory(agent_list[i].traj)
        allcfs.crazyflies[i].uploadTrajectory(0, 0, traj)

    for i in range(Num):  
        allcfs.crazyflies[i].startTrajectory(0,relative=False)
```

Therefore, how to start a correct trajectory synchronously for several crazyfile?

**Comment by whoenig** [comment]
What does `.input_trajectory` do in this case - this is non-standard? Also, you should create the `traj` object inside your first for-loop, see https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/scripts/swap6v.py#L36-L52 for an example. Finally, in your first code snipped, you use `agent[i]`, while the second has `agent_list[i]` - is that intentional?

**Comment by CYDXYYJ** [comment]
`.input_trajectory` is a function defined by myself as a member of `class Trajectory` in `uav_trajectory.py`  as following :
```
  def input_trajectory(self, data):
    self.polynomials = [Polynomial4D(row[0],row[1],row[2],row[3],row[4]) for row in data]
    self.duration = 0
    for i in range(len(data)):
      self.duration+=data[i][0]
```
It is similar to `def loadcsv(self, filename):`.
`agent[i].traj` is the trajectory data like` figure8.csv` in your demo. 

`agent_list[i]` and `agent[i]` are the same variables, they have no difference.

**Comment by whoenig** [comment]
Your two code versions should be equivalent. The only problem I see is that you use `allcfs.crazyflies[i]` rather than `allcfs.crazyfliesById[i]`. The former is an array that has potentially a random order (depending on how you used chooser.py last). The second one is a dictionary and probably what you want in this case.

**Comment by CYDXYYJ** [comment]
You say I should input trajectory by `allcfs.crazyfliesById[i].uploadTrajectory(0, 0, traj)`?

**Reply by whoenig** [reply]
Yes, otherwise you may get a random order which is not what you intended.
